spring-webflux/src/main/kotlin/org/springframework/web/reactive/function/server/ServerResponseExtensions.kt

```diff
@@ -54,14 +55,69 @@ inline fun <reified T : Any> ServerResponse.BodyBuilder.body(producer: Any): Mon
  * Set the body of the response to the given {@code Object} and return it.
  * This convenience method combines [body] and
  * [org.springframework.web.reactive.function.BodyInserters.fromValue].
+ *
+ * ### Deprecation
+ *
+ * This extension is deprecated because it is subject to type erasure.
+ * If an application uses for serialization the `kotlinx.serialization` library and the provided
+ * body uses generics, then the library is unable to serialize the body and spring falls back
+ * to default one. This could be confusing to those who expect data types with generics to be serialized properly
+ * and can lead to unexpected behaviors.
+ *
+ * For example:
+ * ```kotlin
+ * @Serializable
+ * data class SomeBody(@SerialName("user_id") val userId: Int, val name: String)
+ *
+ * suspend fun returnBody(request: ServerRequest): ServerResponse {
+ *     return ServerResponse
+ *         .ok()
+ *         // `fun ServerResponse.BodyBuilder.bodyValueAndAwait(body: Any)`
+ *         // is not a reified function and is subject to type erasure.
+ *         // In cases where body is a collection, the object is serialized as `Collection<Any>`,
+ *         // and the response does not have the proper format, which is not expected.
+ *         // Actual response: [{"userId":1,"name":"name"}]
+ *         // Expected response : [{"user_id":1,"name":"name"}]
+ *         .bodyValueAndAwait(listOf(SomeBody(1, "name")))
+ * }
+ * ```
+ *
  * @param body the body of the response
  * @return the built response
  * @throws IllegalArgumentException if `body` is a [Publisher] or an
- * instance of a type supported by [org.springframework.core.ReactiveAdapterRegistry.getSharedInstance],
+ * instance of a type supported by [org.springframework.core.ReactiveAdapterRegistry.getSharedInstance].
  */
+@Suppress("DeprecatedCallableAddReplaceWith")
+@Deprecated(
+	message = "Use `bodyValueAndAwait<T>()` instead, which is not subject to type erasure.",
+	level = DeprecationLevel.WARNING
+)
+@JvmName("bodyValueAndAwait0") // to avoid platform declaration clash
```

Comment:
Can you provide more details on the issue you see if you remove this `@JvmName("bodyValueAndAwait0")` annotation?

Reply:
If you remove this annotation then the compilation will fail with error:
```text
Platform declaration clash: The following declarations have the same JVM signature (bodyValueAndAwait(Lorg/springframework/web/reactive/function/server/ServerResponse$BodyBuilder;Ljava/lang/Object;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;):
    suspend inline fun <reified T : Any> ServerResponse.BodyBuilder.bodyValueAndAwait(body: T, `$completion`: Continuation<ServerResponse>): Any? defined in org.springframework.web.reactive.function.server
    suspend fun ServerResponse.BodyBuilder.bodyValueAndAwait(body: Any, `$completion`: Continuation<ServerResponse>): Any? defined in org.springframework.web.reactive.function.server
```

When the bytecode is generated the generics are erased so the compiler just sees two functions with the same type, because in the end `T` is replaced with `Any`.